src/jumps/jumpTracker.ts

```diff
@@ -0,0 +1,265 @@
+   * @param from - File/position jumped from
+   */
+  public jumpForward(from: Jump): Jump {
+    if (!this.hasJumps) {
+      return from;
+    }
+
+    if (this._currentJumpNumber >= this._jumps.length) {
+      return from;
+    }
+
+    this._currentJumpNumber = Math.min(this._currentJumpNumber + 1, this._jumps.length - 1);
+    const jump = this._jumps[this._currentJumpNumber];
+    return jump;
+  }
+
+  /**
+   * Update existing jumps when lines were added to a document.
+   *
+   * @param document - Document that was changed.
+   * @param range - Location where the text was added.
+   * @param text - Text containing one or more newline characters.
+   */
+  public handleTextAdded(document: vscode.TextDocument, range: vscode.Range, text: string): void {
+    const distance = text.split('').filter(c => c === '\n').length;
+
+    this._jumps.forEach((jump, i) => {
+      const jumpIsAfterAddedText =
+        jump.fileName === document.fileName && jump.position.line > range.start.line;
+
+      if (jumpIsAfterAddedText) {
+        const newPosition = new Position(jump.position.line + distance, jump.position.character);
+
+        this.changePositionForJumpNumber(i, jump, newPosition);
+      }
+    });
+  }
+
+  /**
+   * Update existing jumps when lines were removed from a document.
+   *
+   * @param document - Document that was changed.
+   * @param range - Location where the text was removed.
+   */
+  public handleTextDeleted(document: vscode.TextDocument, range: vscode.Range): void {
+    const distance = range.end.line - range.start.line;
+
+    this._jumps.forEach((jump, i) => {
+      const jumpIsAfterDeletedText =
+        jump.fileName === document.fileName && jump.position.line >= range.start.line;
+
+      if (jumpIsAfterDeletedText) {
+        const newLineShiftedUp =
+          jump.position.line - Math.min(jump.position.line - range.start.line, distance);
+        const newPosition = new Position(newLineShiftedUp, jump.position.character);
+
+        this.changePositionForJumpNumber(i, jump, newPosition);
+      }
+    });
+
+    this.removeDuplicateJumps();
+  }
+
+  /**
+   * Clear existing jumps and reset jump position.
+   */
+  public clearJumps(): void {
+    this._jumps.splice(0, this._jumps.length);
+    this._currentJumpNumber = 0;
+  }
+
+  changePositionForJumpNumber(index: number, jump: Jump, newPosition: Position) {
+    this._jumps.splice(
+      index,
+      1,
+      new Jump({
+        editor: jump.editor,
+        fileName: jump.fileName,
+        position: newPosition,
+      })
+    );
+  }
+
+  clearOldJumps(): void {
+    if (this._jumps.length > 100) {
+      this._jumps.splice(0, this._jumps.length - 100);
+    }
+  }
+
+  clearJumpsOnSameLine(jump: Jump): void {
+    this._jumps = this._jumps.filter(j => j === jump || !j.onSameLine(jump));
+  }
+
+  removeDuplicateJumps() {
+    for (let i = 0; i < this._jumps.length; i++) {
+      const jump = this._jumps[i];
+      this.clearJumpsOnSameLine(jump);
```

Comment:
Are we concerned about potential N^2 complexity here?

The max N=100, but 100^2 is still 10k, which is bordering on significant.

If any of the comparisons are expensive, then this might become an issue.

Reply:
Also, is it an issue which order we remove jumps?

Reply:
Good points. The latest jump should be preserved, to have the latest column number.

I can loop through the jumps in the list in reverse and keep track of the line numbers as I go, removing when I see a duplicate. That should get it down to O(n) at least and preserve the newer jump.

Reply:
I changed it as described, and I added a new "Can handle text deleted from a file" test to `JumpTracker.tests.js` to validate behavior. Added tests, also added a test for Visual Mode instead of just Visual Line Mode.